sql/src/main/java/io/crate/analyze/relations/RelationSplitter.java

```diff
@@ -172,64 +167,95 @@ private void processOrderBy() {
             return;
         }
         OrderBy orderBy = querySpec.orderBy().get();
+
         Set<AnalyzedRelation> relations = Collections.newSetFromMap(new IdentityHashMap<AnalyzedRelation, Boolean>());
-        Consumer<Field> addToRelations = f -> relations.add(f.relation());
-        Multimap<AnalyzedRelation, Integer> splits = Multimaps.newSetMultimap(
-            new IdentityHashMap<AnalyzedRelation, Collection<Integer>>(specs.size()),
-            INT_SET_SUPPLIER);
+        Consumer<Field> relationsConsumer = f -> relations.add(f.relation());
+
+        // We copy all ORDER BY symbols to remainingOrderBy
+        remainingOrderBy = new RemainingOrderBy();
+        extractRemainingOrderBy(orderBy, relations, relationsConsumer);
 
-        // Detect remaining orderBy before any push down happens,
-        // since if remaining orderBy is detected we need to
-        // process again all pushed down orderBys and merge them
-        // to the remaining OrderBy in the correct order.
+        Multimap<AnalyzedRelation, Integer> splits = extractOrderByForPushDown(orderBy, relations, relationsConsumer);
+
+        // Pushed down the orderBy to subquery specs and also set the limitAndOffset if present
+        Optional<Symbol> limitAndOffset = Limits.mergeAdd(querySpec.limit(), querySpec.offset());
+        for (Map.Entry<AnalyzedRelation, Collection<Integer>> entry : splits.asMap().entrySet()) {
+            AnalyzedRelation relation = entry.getKey();
+            OrderBy newOrderBy = orderBy.subset(entry.getValue());
+            QuerySpec spec = getSpec(relation);
+            assert !spec.orderBy().isPresent() : "spec.orderBy() must not be present";
+            spec.orderBy(newOrderBy);
+            if (limitAndOffset.isPresent()) {
+                spec.limit(limitAndOffset);
+            }
+            requiredForQuery.addAll(newOrderBy.orderBySymbols());
+        }
+    }
+
+    private void extractRemainingOrderBy(OrderBy orderBy, Set<AnalyzedRelation> relations, Consumer<Field> relationsConsumer) {
+        Integer idx = 0;
         for (Symbol symbol : orderBy.orderBySymbols()) {
+            // If order by is pointing to an aggregation we cannot push it down
+            // because ordering needs to happen AFTER the join
+            if (Aggregations.containsAggregation(symbol)) {
+                continue;
```

Comment:
the remainingorderby could be created in this method body, however i am not sure if it is needed when no order symbols are added, which leads me to the point that this continue here is somehow odd, since this would leave the remainings empty?